### PR TITLE
{chem}[intel/2017b] Dalton 2016

### PR DESCRIPTION
**Pull request by hajgato**
All (non-benchmark) test jobs were passed
Slightly different approach compared to #5778

### CODE REVIEW

**Comment by boegel** [comment]
@akesandgren Are you up for giving this a quick review?

@hajgato Can you point out the differences with #5778?

File: easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+]
+
+builddependencies = [('CMake', '3.9.5')]
+
+separate_build_dir = True
+
+configopts = '-DCMAKE_BUILD_TYPE=RELEASE '
+configopts += '-DENABLE_MPI=ON '
+configopts += '-DENABLE_OMP=ON '
+configopts += '-DENABLE_64BIT_INTEGERS=ON '
```
Comment (akesandgren) [inline easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb]: I'd avoid using 64bit ints for the uploaded recipy.
It's not compatible with most scalapack/blacs/lapack/blas builds.

**Review by akesandgren — DISMISSED** [review]

File: easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+configopts += '-DENABLE_AUTO_BLAS=OFF '
+configopts += '-DENABLE_AUTO_LAPACK=OFF '
+configopts += '-DEXPLICIT_BLAS_LIB="$LIBBLAS" '
+configopts += '-DEXPLICIT_LAPACK_LIB="$LIBLAPACK" '
```
Comment (akesandgren) [inline easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb]: You probably need
configopts += '-DBLACS_IMPLEMENTATION=intelmpi '
configopts += '-DENABLE_SCALAPACK=ON'

Not using scalapack will make it a fair bit slower.

**Comment by hajgato** [comment]
Differences compared to #5778 

- 64-bit integers (I would stick to them, will give a comment only use in case of MKL)
- Dalton compiler flags are controlled by `EasyBuild` instead using hard-coded optimization flag `-O3` (although extensions still use hard-coded ones)
- All broken test jobs are fixed (switching back to lowopt `-O1`) when it is needed.

I will fix `ScaLAPACK` as well, but I have to patch `CMake` files, because by default (see switches above) will use the Fortran bundled `ScsLAPACK` instead of the `MKL` one. Anyway, its only beneficial to use for internode MPI jobs.

File: easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb (hunk 2 above)
Comment (hajgato) [reply easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb]: If you check the `CMake` files, you will see that `ScaLAPACK`/`BLACS` only works with `MKL`. Meanwhile enable 64 bit integers would work with `MKL`, `ACLM` out-of-the box, and you can easily make 64 bit integer version of `OPENBLAS` , and `ATLAS`.  So I think the same should apply to `ENABLE_64BIT_INTEGERS=ON` and `ENABLE_SCALAPACK=ON`

**Comment by hajgato** [comment]
I was wrong, `CMake` will actually use `MKL` from the `imkl` not the bundled one.

**Comment by hajgato** [comment]
@akesandgren @boegel ^

**Comment by migueldiascosta** [comment]
@akesandgren are you ok with the scalapack changes and added comment on 64bit integers? (or maybe this one should have an `_i8` versionsuffix?)

File: easybuild/easyconfigs/d/Dalton/Dalton-2016_cmake-obey-EB-variables.patch [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ 
+diff -ru dalton-release-2016-130ffaa0613bb3af6cac766fc8183d6df7d68917.orig/cmake/compilers/FortranFlags.cmake dalton-release-2016-130ffaa0613bb3af6cac766fc8183d6df7d68917/cmake/compilers/FortranFlags.cmake
+--- dalton-release-2016-130ffaa0613bb3af6cac766fc8183d6df7d68917.orig/cmake/compilers/FortranFlags.cmake	2017-08-14 15:00:21.000000000 +0200
++++ dalton-release-2016-130ffaa0613bb3af6cac766fc8183d6df7d68917/cmake/compilers/FortranFlags.cmake	2018-02-06 14:50:21.404219685 +0100
+@@ -48,13 +48,13 @@
+ 
+ if(CMAKE_Fortran_COMPILER_ID MATCHES Intel)
+     add_definitions(-DVAR_IFORT)
+-    set(CMAKE_Fortran_FLAGS         "-fpp -assume byterecl -DVAR_IFORT")
++    set(CMAKE_Fortran_FLAGS         "${CMAKE_Fortran_FLAGS} -fpp -DVAR_IFORT")
```
Comment (akesandgren) [inline easybuild/easyconfigs/d/Dalton/Dalton-2016_cmake-obey-EB-variables.patch]: Missed this one last time around.
I'm fairly certain you have to keep -assume byterecl for Intel fortran

**Comment by akesandgren** [comment]
It might be worth it to have the -i8 build have a versionsuffix specifying this, but since -i8 is the preferred way for dalton, one might set versionsuffix -i4 on the non-i8 builds...
Don't care one way or the other, as long as they are clearly distinct.

And -i8/4 is probably not the most informative suffix, but I don't know of any better at the moment.

At the very least one should add some explanation to the module file.

We never bothered building it with -i8 since there was never a reason to do it for our users.

Other than that I have no objections to it.

File: easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    '4942d95e77845fab8b196c88f390e630971559e51632ca280434dbc0eed26757',
+]
+
+builddependencies = [('CMake', '3.9.5')]
+
+separate_build_dir = True
+
+configopts = '-DCMAKE_BUILD_TYPE=RELEASE '
+configopts += '-DENABLE_MPI=ON '
+configopts += '-DENABLE_OMP=ON '
```
Comment (akesandgren) [inline easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb]: And if you enable OMP you should probably add openmp: True in toolchainopts?

File: easybuild/easyconfigs/d/Dalton/Dalton-2016_cmake-obey-EB-variables.patch (hunk 3 above)
Comment (hajgato) [reply easybuild/easyconfigs/d/Dalton/Dalton-2016_cmake-obey-EB-variables.patch]: @akesandgren Not necessarily, the `DIRECT_IO_FACTOR` can be 1 or 4 with or without byterecl. There is a setting for `Dalton` I forget where it is, i vaguely remember that. Anyway, if you use the wring model, `Dalton will not work at all.

File: easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb (hunk 4 above)
Comment (hajgato) [reply easybuild/easyconfigs/d/Dalton/Dalton-2016-intel-2017b.eb]: Done

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2447.golett.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/e84419eceb392362d2df5b8f6673914a for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3404.kirlia.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 2.7.5
See https://gist.github.com/09be0a71edb08611cacaec59675d0625 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3134.skitty.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/20c15b4cc1a7e129d2715394400c7973 for a full test report.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Going in, thanks @hajgato!

**Pull request merged**